modules/images/regenerate-existing-images/background-process/class-perflab-background-job.php

```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @since n.e.x.t
+	 * @var string Job status failed.
+	 */
+	const JOB_STATUS_FAILED = 'perflab_job_failed';
+
+	/**
+	 * Job ID.
+	 *
+	 * @since n.e.x.t
+	 * @var int
+	 */
+	public $job_id;
+
+	/**
+	 * Job name.
+	 *
+	 * @since n.e.x.t
+	 * @var string Job name.
+	 */
+	private $name;
+
+	/**
+	 * Job data.
+	 *
+	 * @since n.e.x.t
+	 * @var array Job data.
+	 */
+	private $data;
+
+	/**
+	 * Perflab_Background_Job constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $job_id job ID.
+	 */
+	public function __construct( $job_id = 0 ) {
+		$this->job_id = absint( $job_id );
+
+		if ( $this->exists() ) {
+			$this->name = get_term_meta( $this->job_id, self::JOB_NAME_META_KEY, true );
+			$this->data = get_term_meta( $this->job_id, self::JOB_DATA_META_KEY, true );
+		}
+	}
+
+	/**
+	 * Determines whether a job should run for current request or not.
+	 *
+	 * It determines by checking if job is already running or completed.
+	 * It also checks if max number of attempts have been attempted for the job.
+	 * For any such condition, it will return false, else true.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return bool Flag to tell whether this job should run or not.
+	 */
+	public function should_run() {
+		// If job doesn't exist or completed, return false.
+		if ( ! $this->exists() || $this->is_completed() ) {
+			return false;
+		}
+
+		if ( $this->is_running() ) {
+			return false;
+		}
+
+		/**
+		 * Number of attempts to try to run a job.
+		 * Repeated attempts may be required to run a failed job. Default 3 attempts.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param int $attempts Number of attempts allowed for a job to run.
+		 */
+		$max_attempts = apply_filters( 'perflab_job_max_attempts_allowed', 3 );
+
+		// If number of attempts have been exhausted, return false.
+		if ( $this->get_attempts() >= $max_attempts ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Sets the status of the current job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $status Status to set for current job.
+	 *
+	 * @return bool Returns true if status is updated, false otherwise.
+	 */
+	public function set_status( $status ) {
+		$valid_statuses = array(
+			self::JOB_STATUS_COMPLETE,
+			self::JOB_STATUS_FAILED,
+			self::JOB_STATUS_QUEUED,
+			self::JOB_STATUS_RUNNING,
+			self::JOB_STATUS_PARTIAL,
+		);
+
+		if ( in_array( $status, $valid_statuses, true ) ) {
+			update_term_meta( $this->job_id, self::JOB_STATUS_META_KEY, $status );
+
+			// If job is complete, set the timestamp at which it was completed.
+			if ( self::JOB_STATUS_COMPLETE === $status ) {
+				update_term_meta( $this->job_id, 'job_completed_at', time() );
+			}
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Records the error for the current process run.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param WP_Error $error Error instance.
+	 *
+	 * @todo Implement the error recording activity.
+	 *
+	 * @return void
+	 */
+	public function set_error( WP_Error $error ) {
```

Comment:
@ankitrox This works differently to what's defined in #485. However, I think we can readdress this when working on this issue which will need updating based on recent addition of the Job class.

Reply:
@jjgrainger I have left it like this intentionally to handle this in different PR. We also have a set format for error data and other things, so we can handle this separately as you mentioned.